### PR TITLE
[libraries/User] Add checks and error logging to prevent notices and warnings

### PR DESCRIPTION
**Pull request by johnsaigle**
- [x] Ensure array values exist before passing them into SQL queries

- [x] Provide more verbose error logging to assist developers in debugging 

- [x] Prevent generation of vague PHP Notices when these values did not exist but were used 

**Note**: These issues largely came up while using `request_account` so it's possible that some of the execution flow in that code is responsible (e.g. looking for examiner associations before the user is actually created?). As such, the underlying issue may exist somewhere other than this class.

Still, these checks should exist and if the issues lie in other places in the code base, this will aid us in discovering and fixing them.

### CODE REVIEW

File: php/libraries/User.class.inc [hunk 1]
```diff
@@ -66,26 +66,41 @@ class User extends UserPermissions
         $row = $DB->pselectRow($query, array('UID' => $username));
```
Comment (davidblader) [inline php/libraries/User.class.inc]: should there be a check to see if $username is actually set before setting / retrieving data above?

Comment (johnsaigle) [inline php/libraries/User.class.inc]: The function is called using `$username` so I don't think it's possible for it to be unset

Comment (davidblader) [inline php/libraries/User.class.inc]: sure, but i don't get any warning or error for if User::factory() is called without an argument. is that normal?

Comment (johnsaigle) [inline php/libraries/User.class.inc]: Yeah good point. I think it's possible for that situation to arise in `request_account` which is exactly how I stumbled on this. I'll add another error message.

**Comment by johnsaigle** [comment]
@davidblader I added a check to throw an exception when the `$username` is null.

**Comment by xlecours** [comment]
I have rebased/merge minor into it in https://github.com/aces/Loris/pull/3196

**Pull request closed without merging**